### DISCUSSION TITLE
Sending MIDI Show Control Messages

### DISCUSSION THREAD

**Discussion by razvanbalint** [orchetect/swift-midi#163] [Q&A] [answered]
I've noticed sending raw MIDI messages capability is marked as internal (`_MIDIManagedSendsMessages`) and from my understanding only sending the predefined `MIDIEvent`s is possible. 
I'm actually interested in sending MIDI Show Control commands and therefore I need to be able to create specific format MIDI messages.
Is this possible using this library?

**Comment by orchetect** [comment, marked as answer]
All possible valid MIDI messages can be constructed using `MIDIEvent`. MIDI Show Control uses standard SysEx7 messages.

The [MIDI Show Control Spec](https://www.midi.org/specifications/midi1-specifications/midi-show-control-2) says:

> MIDI Show Control uses a single Universal Real Time System Exclusive ID number (sub-ID # 1 = 0x02) for all Show commands (transmissions from Controller to Controlled Device).
>
> The format of a Show Control message is as follows:
> `F0 7F <device_ID> <msc> <command_format> <command> <data> F7`

There's actually three options that can work for you. The following examples all construct exactly the same message.

1. The [MIDIEvent.sysEx7(rawBytes:)](https://orchetect.github.io/MIDIKit/documentation/midikit/midievent/sysex7(rawbytes:group:)) event constructor will take an array of MIDI 1.0 SysEx bytes. It only throws an error if the SysEx message is malformed.

   ```swift
   let event: MIDIEvent = try .sysEx7(rawBytes: [0xF0, 0x7F, 0x01, 0x02, 0x01, 0x07, 0x0A, 0xF7])
   ```

2. The [MIDIEvent.sysEx7(rawHexString:)](https://orchetect.github.io/MIDIKit/documentation/midikit/midievent/sysex7(rawhexstring:group:)) constructor does the same thing, but conveniently takes a hex byte string. Hex bytes can be contiguous or padded with a space between them. It only throws an error if the string is not valid hex bytes, or if the SysEx message is malformed.

   ```swift
   let event: MIDIEvent = try .sysEx7(rawHexString: "F07F010201070AF7")
   let event: MIDIEvent = try .sysEx7(rawHexString: "F0 7F 01 02 01 07 0A F7")
   ```

3. The [MIDIEvent.sysEx7.universalSysEx7](https://orchetect.github.io/MIDIKit/documentation/midikit/midievent/universalsysex7(universaltype:deviceid:subid1:subid2:data:group:)) constructor will also work since MIDI Show Control specifically uses Universal SysEx format. It just separates the header parts into more descriptive types, as per the MIDI spec. (With this one, you only supply the inner data bytes, so the F0/F7 trailing SysEx bytes and header info is taken care of for you.)

   ```swift
   let event: MIDIEvent = .universalSysEx7(
       universalType: .realTime, // byte two (0x7F)
       deviceID: 0x01, // byte three (<device_ID>)
       subID1: 0x02, // byte four (<msc>)
       subID2: 0x01, // byte five (<command_format>)
       data: [0x07, 0x0A] // byte six and onward, but do not include trailing 0xF7
   )
   ```

**Comment by razvanbalint** [comment]
I haven't noticed the `sysEx7` methods, my bad. 
Your detailed response is highly appreciated. Keep up the good work!

**Reply by orchetect** [reply]
My pleasure. I am adding all this info to the docs for the benefit of other users too:

https://orchetect.github.io/MIDIKit/documentation/midikit/midi-show-control
